warmup/summary_statistics.py

```diff
@@ -0,0 +1,299 @@
+            # Lists of changepoints, outliers and segment means for each process execution.
+            changepoints, outliers, segments = list(), list(), list()
+            for p_exec in xrange(n_pexecs):
+                changepoints.append(data_dictionaries[machine]['changepoints'][key][p_exec])
+                segments.append(data_dictionaries[machine]['changepoint_means'][key][p_exec])
+                outliers.append(data_dictionaries[machine]['all_outliers'][key][p_exec])
+                categories.append(data_dictionaries[machine]['classifications'][key][p_exec])
+                # Next we calculate the iteration at which a steady state was
+                # reached, it's average segment mean and the time to reach a
+                # steady state. However, the last segment may be equivalent to
+                # its adjacent segments, so we first need to know which segments
+                # are steady-state segments.
+                first_steady_segment = len(data_dictionaries[machine]['changepoint_means'][key][p_exec]) - 1
+                num_steady_segments = 1
+                last_segment_mean = data_dictionaries[machine]['changepoint_means'][key][p_exec][-1]
+                lower_bound = min(last_segment_mean * (1.0 - half_bound), last_segment_mean - delta)
+                upper_bound = max(last_segment_mean * (1.0 + half_bound), last_segment_mean + delta)
+                for index in xrange(len(data_dictionaries[machine]['changepoint_means'][key][p_exec]) - 2, -1, -1):
+                    current_segment_mean = data_dictionaries[machine]['changepoint_means'][key][p_exec][index]
+                    if current_segment_mean >= lower_bound and current_segment_mean <= upper_bound:
+                        first_steady_segment -= 1
+                        num_steady_segments += 1
+                    else:
+                        break
+                steady_state_mean = (math.fsum(data_dictionaries[machine]['changepoint_means'][key][p_exec][first_steady_segment:])
+                                     / float(num_steady_segments))
+                steady_state_means.append(steady_state_mean)
+                # Not all process execs have changepoints. However, all
+                # p_execs will have one or more segment mean.
+                if data_dictionaries[machine]['changepoints'][key][p_exec]:
+                    steady_iter = data_dictionaries[machine]['changepoints'][key][p_exec][first_steady_segment - 1]
+                    steady_iters.append(steady_iter)
+                    to_steady = 0.0
+                    for index in xrange(steady_iter):
+                        to_steady += data_dictionaries[machine]['wallclock_times'][key][p_exec][index]
+                    time_to_steadys.append(to_steady)
+                else:  # Flat execution, no changepoints.
+                    time_to_steadys.append(0.0)
+            # Get overall and detailed categories.
```

Comment:
Would it be better to say `categories_set = set(categories)` rather than sometimes calculating it twice?